### ISSUE TITLE
Loading of the pages takes to long

### ISSUE THREAD

**Issue by Marko-ilmkreis** [hzi-braunschweig/SORMAS-Project#9559]
### Bug Description

Since the update from 1.70.4 to 1.72.1 yesterday we see a significant increase of the loading times in Sormas.

I traced the following times to loads the pages:

User Role: admin, national user, national clinician:

Dachboard: **11 sec**
Tasks: 1,16 sec (118 tasks shown)
Persons: 2,26 sec (47142 persons shown for cases, when switching to all persons (84884) it loads only 3,22 sec)
Cases: 1,48 sec (17814 cases)
Contacts: **32,7 sec** (2934 contacts)
Events: **33 sec** (591 events)
Samples: 2,45 sec ( 53000 samples)
Entries: 1,5 sec (no entries in database)

User Role: Clinician, Contact Supervisor, Surveillance Supervisor

Dachboard: **7 sec**
Contacts: **7 sec** 
Events: **8 sec** 

Netzlink is getting the log-files to check them. 

### Screenshots

![image](https://user-images.githubusercontent.com/74791118/174716695-39943e8b-035a-415c-b876-28f22b8b3409.png)

### System Details

* Device:
* SORMAS version: 1.72.1
* Android version/Browser: Edge
* Server URL: own instance
* User Role: admin, national user, national clinician

**Comment by kwa20** [comment]
@Marko-ilmkreis Thanks for uploading this issue! I think the jump in versioning was actually 1.70.4 to 1.72.1 correct?

**Comment by kwa20** [comment]
Tested further with Admin + National User respectively

  | test-performance.sormas.netzlink.com   (1.74-SNAPSHOT(3316ffe)) |   | schnischu.sormas.netzlink.com (1.72.1) |  
-- | -- | -- | -- | --
Test step | time | number of entries | time | number of entries
Login | 10.92 s | NA | 9.49 s | NA
Open Surveillance Dashboard | 10.51 s | Unknown | 8.96 s | Unknown
Open Contact Dashboard | 24.02 s | Unknown | 10.21 s | Unknown
Open task Management | 5.80 s | 241639 | 3.41 s | 103
Open Person Directory | 16.61 s | 61202 | 1.27 s | 676
Filter for all persons | 1.1 min | 1041498 | 1.20 s | 1124
Open a person | 5.72 s | NA | 4.90 s | NA
Open Cases Directory | 2.12 s | 67479 | 1.44 s | 305
Filter for archived cases | 659 ms | 0 | 1.36 s | 427
Filter for active cases | 1.28 s | 67479 | 1.34 s | 305
Open a case | 10.17 s | NA | 8.92 s | NA
Open Contacts Directory | 7.99 s | 2038 | 54.79 s | 280
Filter for archived contacts | 8.25 s | 116950 | 1.8 min | 239
Filter for active contacts | 7.47 s | 2038 | 34.08 s | 280
Open a contact | 6.13 s | NA | 11.59 s | NA
Open Events Directory | 4.55 s | 14051 | 4.92 s | 10
Filter for archived events | 594 ms | 0 | 27.61 s | 73
Filter for active events | 27.69 s | 14051 | 4.76 s | 10
Open an event | 4.91 s | NA | 5.29 s | NA
Open Samples Directory | 1.17 s | 12123 | 1.55 s | 192
Filter for archived samples | 957 ms | 0 | 1.53 s | 148
Filter for active samples | 1.09 s | 12123 | 1.60 s | 192
Open a sample | 3.84 s | NA | 3.93 s | NA
Open the Immunizations Directory | 13.43 s | Unknown | NA | NA
Open the Travel Entries Directory | NA | NA | 1.38 s | 0

**Comment by syntakker** [comment]
Removing the `@RolesAllowed(UserRight._TASK_VIEW)` annotation in `TaskFacadeEjb` for test purposes lets disappear the latency between call of the EJB method in `EventGrid` and the method actually being called.

<details>
<summary>log extract</summary>

without annotation:
```diff
2022-06-22 04:50:30,601 DEBUG NatiUser  http-thread-pool::http-listener-1(2) de.symeda.sormas.ui.events.EventGrid - Start getTaskFacade
2022-06-22 04:50:30,603 DEBUG NatiUser  http-thread-pool::http-listener-1(2) de.symeda.sormas.ui.events.EventGrid - Start getPendingTaskCountByEvent
----> no gap
2022-06-22 04:50:30,603 TRACE NatiUser  http-thread-pool::http-listener-1(2) d.s.s.b.u.PerformanceLoggingInterceptor - Started: TaskFacadeEjb.getPendingTaskCountByEvent with parameters '[ Signal Untersuchung ausstehend an 26.01.2022]'
2022-06-22 04:50:30,604 DEBUG NatiUser  http-thread-pool::http-listener-1(2) d.s.s.backend.task.TaskFacadeEjb - Start getPendingTaskCountByEvent
2022-06-22 04:50:30,604 TRACE NatiUser  http-thread-pool::http-listener-1(2) d.s.s.b.u.PerformanceLoggingInterceptor - Started: TaskService.getCount with parameters '[de.symeda.sormas.api.task.TaskCriteria@ae5bb0]'
2022-06-22 04:50:30,605 DEBUG NatiUser  http-thread-pool::http-listener-1(2) d.s.s.b.u.PerformanceLoggingInterceptor - Finished in 1 ms: TaskService.getCount
2022-06-22 04:50:30,606 DEBUG NatiUser  http-thread-pool::http-listener-1(2) d.s.s.b.u.PerformanceLoggingInterceptor - Finished in 2 ms: TaskFacadeEjb.getPendingTaskCountByEvent
2022-06-22 04:50:30,606 DEBUG NatiUser  http-thread-pool::http-listener-1(2) de.symeda.sormas.ui.events.EventGrid - Finish getPendingTaskCountByEvent
```

with annotation:
```diff
2022-06-22 05:06:32,314 DEBUG NatiUser  http-thread-pool::http-listener-1(1) de.symeda.sormas.ui.events.EventGrid - Start getTaskFacade
2022-06-22 05:06:32,316 DEBUG NatiUser  http-thread-pool::http-listener-1(1) de.symeda.sormas.ui.events.EventGrid - Start getPendingTaskCountByEvent
----> gap: 200ms
2022-06-22 05:06:32,516 TRACE NatiUser  http-thread-pool::http-listener-1(1) d.s.s.b.u.PerformanceLoggingInterceptor - Started: TaskFacadeEjb.getPendingTaskCountByEvent with parameters '[ Signal Untersuchung ausstehend an 26.01.2022]'
2022-06-22 05:06:32,517 DEBUG NatiUser  http-thread-pool::http-listener-1(1) d.s.s.backend.task.TaskFacadeEjb - Start getPendingTaskCountByEvent
2022-06-22 05:06:32,518 TRACE NatiUser  http-thread-pool::http-listener-1(1) d.s.s.b.u.PerformanceLoggingInterceptor - Started: TaskService.getCount with parameters '[de.symeda.sormas.api.task.TaskCriteria@30aedbcf]'
2022-06-22 05:06:32,520 DEBUG NatiUser  http-thread-pool::http-listener-1(1) d.s.s.b.u.PerformanceLoggingInterceptor - Finished in 2 ms: TaskService.getCount
2022-06-22 05:06:32,520 DEBUG NatiUser  http-thread-pool::http-listener-1(1) d.s.s.b.u.PerformanceLoggingInterceptor - Finished in 3 ms: TaskFacadeEjb.getPendingTaskCountByEvent
2022-06-22 05:06:32,520 DEBUG NatiUser  http-thread-pool::http-listener-1(1) de.symeda.sormas.ui.events.EventGrid - Finish getPendingTaskCountByEvent
```
</details>

**Comment by StefanKock** [comment]
Next steps:
1. @MartinWahnschaffe Cross check if adding `@RolesAllowed` to `DiseaseConfigurationFacadeEjb` also then gets delayed.
2. Start working on #9569.
3. PoC: Replace `@RolesAllowed` with own implementation that replicated its behaviour together with an Interceptor.
4. Debug in the Jakarta EE & Payara stack where the roles check happens and causes delays.

**Comment by MartinWahnschaffe** [comment]
Adding `@RolesAllowed` to `DiseaseConfigurationFacadeEjb` also leads to the delay.  I tried this both with the TASKS_VIEW and a random other user right (SORMAS_UI) - same result.

**Comment by StefanKock** [comment]
I activated the logging for [EJBSecurityManager](https://github.com/payara/Payara/blob/payara-server-5.2021.10/appserver/ejb/ejb-container/src/main/java/org/glassfish/ejb/security/application/EJBSecurityManager.java) (`config/logging.properties` -> `javax.enterprise.system.container.ejb.level=FINE`, worked with running domain for me).

<details>
<summary>Log with EJBSecurityManager</summary>

```

[2022-06-22T12:21:50.639+0200] [Payara 5.2021.10] [FEIN] [] [javax.enterprise.system.container.ejb.org.glassfish.ejb.security.application] [tid: _ThreadID=111 _ThreadName=http-thread-pool::http-listener-1(3)] [timeMillis: 1655893310639] [levelValue: 500] [CLASSNAME: org.glassfish.ejb.security.application.EJBSecurityManager] [METHODNAME: resetPolicyContext] [[
  JACC: Changing Policy Context ID: oldV = null newV = sormas-ear/sormas-backend_jar]]

[2022-06-22T12:21:50.639+0200] [Payara 5.2021.10] [FEIN] [] [javax.enterprise.system.container.ejb.org.glassfish.ejb.security.application] [tid: _ThreadID=111 _ThreadName=http-thread-pool::http-listener-1(3)] [timeMillis: 1655893310639] [levelValue: 500] [CLASSNAME: org.glassfish.ejb.security.application.EJBSecurityManager] [METHODNAME: getCachedProtectionDomain] [[
  JACC: new ProtectionDomain added to cache]]

[2022-06-22T12:21:50.640+0200] [Payara 5.2021.10] [FEIN] [] [javax.enterprise.system.container.ejb.org.glassfish.ejb.security.application] [tid: _ThreadID=111 _ThreadName=http-thread-pool::http-listener-1(3)] [timeMillis: 1655893310640] [levelValue: 500] [CLASSNAME: org.glassfish.ejb.security.application.EJBSecurityManager] [METHODNAME: getCachedProtectionDomain] [[
  JACC: returning cached ProtectionDomain - CodeSource: ((file:/sormas-ear/sormas-backend_jar <no signer certificates>)) PrincipalSet: javax.security.enterprise.CallerPrincipal@7f7eddd7 ADDITIONAL_TEST_VIEW EXPORT_DATA_PROTECTION_DATA DASHBOARD_CAMPAIGNS_VIEW EVENTPARTICIPANT_CREATE TASK_ASSIGN IMMUNIZATION_VIEW POPULATION_MANAGE PERSON_EDIT ACTION_EDIT CASE_ARCHIVE SORMAS_UI DOCUMENT_TEMPLATE_MANAGEMENT TRAVEL_ENTRY_EDIT CLINICAL_COURSE_EDIT VISIT_EXPORT EXTERNAL_MESSAGE_PROCESS VISIT_CREATE EVENT_VIEW PERSON_EXPORT VISIT_EDIT PATHOGEN_TEST_DELETE CAMPAIGN_FORM_DATA_ARCHIVE CAMPAIGN_DELETE TASK_EDIT CASE_CLASSIFY EVENTPARTICIPANT_VIEW AGGREGATE_REPORT_VIEW CASE_CREATE CONTACT_MERGE STATISTICS_EXPORT SAMPLE_VIEW CONTACT_ASSIGN DATABASE_EXPORT_ACCESS EVENTGROUP_ARCHIVE ADDITIONAL_TEST_DELETE TRAVEL_ENTRY_CREATE EVENTGROUP_DELETE CASE_EXPORT EVENTPARTICIPANT_ARCHIVE ACTION_DELETE CONTACT_CLASSIFY INFRASTRUCTURE_EXPORT INFRASTRUCTURE_VIEW CASE_CHANGE_EPID_NUMBER INFRASTRUCTURE_CREATE QUARANTINE_ORDER_CREATE CASE_INVESTIGATE PRESCRIPTION_CREATE TREATMENT_DELETE USER_CREATE PERSON_CONTACT_DETAILS_DELETE CONTACT_DELETE TRAVEL_ENTRY_ARCHIVE SAMPLE_EDIT DEV_MODE CASE_CHANGE_DISEASE CASE_IMPORT CASE_DELETE CLINICAL_VISIT_CREATE CASE_CLINICIAN_VIEW MANAGE_PUBLIC_EXPORT_CONFIGURATION SAMPLE_DELETE CAMPAIGN_FORM_DATA_EDIT PERFORM_BULK_OPERATIONS_PSEUDONYM MANAGE_EXTERNAL_SYMPTOM_JOURNAL SEE_SENSITIVE_DATA_IN_JURISDICTION CAMPAIGN_VIEW CONTACT_CREATE PORT_HEALTH_INFO_EDIT OUTBREAK_VIEW SAMPLE_EXPORT DASHBOARD_CONTACT_VIEW_TRANSMISSION_CHAINS CASE_MERGE ADDITIONAL_TEST_EDIT PRESCRIPTION_EDIT CASE_EDIT THERAPY_VIEW EXTERNAL_MESSAGE_DELETE EVENT_EXPORT PERFORM_BULK_OPERATIONS_EVENT STATISTICS_ACCESS EVENTGROUP_CREATE TREATMENT_EDIT SAMPLE_EDIT_NOT_OWNED CLINICAL_VISIT_DELETE USER_VIEW CASE_REFER_FROM_POE IMMUNIZATION_CREATE CASE_SHARE CLINICAL_COURSE_VIEW CONTACT_VIEW SAMPLE_CREATE CASE_VIEW CONTACT_EXPORT PERFORM_BULK_OPERATIONS_EVENTPARTICIPANT TRAVEL_ENTRY_VIEW SAMPLE_TRANSFER IMMUNIZATION_EDIT WEEKLYREPORT_VIEW CLINICAL_VISIT_EDIT TRAVEL_ENTRY_DELETE AGGREGATE_REPORT_EDIT TRAVEL_ENTRY_MANAGEMENT_ACCESS CAMPAIGN_FORM_DATA_EXPORT CASE_TRANSFER SEE_PERSONAL_DATA_IN_JURISDICTION SORMAS_REST CONTACT_REASSIGN_CASE EVENT_IMPORT EXTERNAL_MESSAGE_VIEW TASK_EXPORT CAMPAIGN_ARCHIVE INFRASTRUCTURE_ARCHIVE CONTACT_CONVERT EVENT_DELETE TASK_CREATE PERFORM_BULK_OPERATIONS TASK_VIEW VISIT_DELETE PERFORM_BULK_OPERATIONS_CASE_SAMPLES CONTACT_IMPORT ACTION_CREATE DASHBOARD_SURVEILLANCE_VIEW EVENT_CREATE AGGREGATE_REPORT_EXPORT PORT_HEALTH_INFO_VIEW SORMAS_TO_SORMAS_SHARE CAMPAIGN_EDIT LINE_LISTING_CONFIGURE PERSON_DELETE EVENTPARTICIPANT_EDIT IMMUNIZATION_ARCHIVE CAMPAIGN_FORM_DATA_DELETE OUTBREAK_EDIT INFRASTRUCTURE_EDIT EVENTPARTICIPANT_IMPORT EVENTGROUP_EDIT PATHOGEN_TEST_EDIT DASHBOARD_CONTACT_VIEW INFRASTRUCTURE_IMPORT TASK_DELETE EVENT_ARCHIVE IMMUNIZATION_DELETE USER_EDIT EVENTGROUP_LINK CONTACT_EDIT EVENTPARTICIPANT_DELETE PRESCRIPTION_DELETE SEND_MANUAL_EXTERNAL_MESSAGES CONTACT_ARCHIVE TREATMENT_CREATE PERSON_VIEW PERFORM_BULK_OPERATIONS_EXTERNAL_MESSAGES PATHOGEN_TEST_CREATE ADDITIONAL_TEST_CREATE CAMPAIGN_FORM_DATA_VIEW EVENT_EDIT]]

[2022-06-22T12:21:50.870+0200] [Payara 5.2021.10] [FEIN] [] [javax.enterprise.system.container.ejb.org.glassfish.ejb.security.application] [tid: _ThreadID=111 _ThreadName=http-thread-pool::http-listener-1(3)] [timeMillis: 1655893310870] [levelValue: 500] [CLASSNAME: org.glassfish.ejb.security.application.EJBSecurityManager] [METHODNAME: authorize] [[
  JACC: Access Control Decision Result: true EJBMethodPermission (Name) = TaskFacade (Action) = getPendingTaskCountByEvent,Remote,de.symeda.sormas.api.event.EventReferenceDto (Caller) = null]]

[2022-06-22T12:21:50.871+0200] [Payara 5.2021.10] [INFORMATION] [] [] [tid: _ThreadID=111 _ThreadName=http-thread-pool::http-listener-1(3)] [timeMillis: 1655893310871] [levelValue: 800] [[
  T122150.871|     > Started: TaskFacadeEjb.getPendingTaskCountByEvent with parameters '[ Signal Investigation pending on ]' (de.symeda.sormas.backend.util.PerformanceLoggingInterceptor.java:42)
]]

[2022-06-22T12:21:50.871+0200] [Payara 5.2021.10] [FEIN] [] [javax.enterprise.system.container.ejb.org.glassfish.ejb.security.application] [tid: _ThreadID=111 _ThreadName=http-thread-pool::http-listener-1(3)] [timeMillis: 1655893310871] [levelValue: 500] [CLASSNAME: org.glassfish.ejb.security.application.EJBSecurityManager] [METHODNAME: authorize] [[
  JACC: Access Control Decision Result: true EJBMethodPermission (Name) = TaskService (Action) = getCount,Local,de.symeda.sormas.api.task.TaskCriteria (Caller) = null]]

[2022-06-22T12:21:50.871+0200] [Payara 5.2021.10] [INFORMATION] [] [] [tid: _ThreadID=111 _ThreadName=http-thread-pool::http-listener-1(3)] [timeMillis: 1655893310871] [levelValue: 800] [[
  T122150.871|     > Started: TaskService.getCount with parameters '[de.symeda.sormas.api.task.TaskCriteria@6fa435eb]' (de.symeda.sormas.backend.util.PerformanceLoggingInterceptor.java:42)
]]

[2022-06-22T12:21:50.874+0200] [Payara 5.2021.10] [INFORMATION] [] [] [tid: _ThreadID=111 _ThreadName=http-thread-pool::http-listener-1(3)] [timeMillis: 1655893310874] [levelValue: 800] [[
  D122150.874|     > Finished in 2 ms: TaskService.getCount (de.symeda.sormas.backend.util.PerformanceLoggingInterceptor.java:51)
]]

[2022-06-22T12:21:50.875+0200] [Payara 5.2021.10] [INFORMATION] [] [] [tid: _ThreadID=111 _ThreadName=http-thread-pool::http-listener-1(3)] [timeMillis: 1655893310875] [levelValue: 800] [[
  D122150.875|     > Finished in 4 ms: TaskFacadeEjb.getPendingTaskCountByEvent (de.symeda.sormas.backend.util.PerformanceLoggingInterceptor.java:51)
]]

```

</details>

It shows that the delay time happens in `EJBSecurityManager.authorize` after `getCachedProtectionDomain` invoke in line 294. It looks like `policy.implies` causes the problem (subclass of `java.security.Policy`).

**Comment by MartinWahnschaffe** [comment]
Findings so far:

1. The logic calls either EJBSecurityManager.authorize (when using RolesAllowed) or isCallerInRole (when calling SessionContext.isCallerInRole). See https://github.com/payara/Payara/blob/payara-server-5.2021.10/appserver/ejb/ejb-container/src/main/java/org/glassfish/ejb/security/application/EJBSecurityManager.java
2. Within that logic the call to getCachedProtectionDomain always returns a new instance, because the cache is cleared when calling setPolicyContext a few lines afterwards. 
3. This leads to the call to policy.implies not being able to use the ProtectionDomainCache (see sun.security.provider.PolicyFile.implies). The new call to getPermissions is very expensive, because there are 5000+ entries for all public methods of the facades.
4. So the main problem seems to be that PolicyContext.getContextID() returns null, because something else is resetting the context id to null before the whole stuff is called.

It's not clear whether this behavior is wanted or related to the configuration of the server. As far as I can tell we are not doing anything special here, so it may simply be bad design.

This should go into new tickets

- [ ] In order to validate this, we have to create an "empty" jee backend test project and deploy it into a default payara server. Then debug whether the behavior is the same. If so, report in payara forums and ask for help.
- [x] Create a workaround that doesn't use isCallerInRole of the session -> #9577

**Comment by StefanKock** [comment]
Currently no active work because we focus on #9577.

**Comment by MartinWahnschaffe** [comment]
As discussed with  we will not create the investigation ticket. 
The result of the investigation would be one of:

1. There is a bug in payara that is hopefully fixed at some point
2. There is something about our configuration/setup that we need to change to make it properly work
3. There is nothing we can do, because it's working the way it's supposed and can't be configured differently

Identifying the exact problem will very likely take a lot of time (workdays to -weeks). The benefit of doing the investigation would be to potentially come out with result 1 or 2 which would allow us to use the Jakarta EE mechanics for this.
On the other hand the solution that was implemented in #9577 is very straightforward, having only 20ish lines of logic and is a full replacement of the functionality needed by us.
All in all the small benefit of going back to the Jakarta implementation is not worth the effort, so we will not invest more time on this.

There are still three usages of RolesAllowed in sormas-rest that will be replaced as port of #8934
